src/Microsoft.Azure.ServiceBus/ServiceBusDiagnosticsSource.cs

```diff
@@ -0,0 +1,773 @@
+            {
+                message.UserProperties[ActivityIdPropertyName] = id;
+                if (correlationContext != null)
+                {
+                    message.UserProperties[CorrelationContextPropertyName] = correlationContext;
+                }
+            }
+        }
+
+        private string SerializeCorrelationContext(IList<KeyValuePair<string,string>> baggage)
+        {
+            if (baggage.Any())
+            {
+                return string.Join(",", baggage.Select(kvp => kvp.Key + "=" + kvp.Value));
+            }
+            return null;
+        }
+
+        private void SetRelatedOperations(Activity activity, IList<Message> messageList)
+        {
+            if (messageList != null && messageList.Count > 0)
+            {
+                var relatedTo = new List<string>();
+                foreach (var message in messageList)
+                {
+                    if (message.TryExtractId(out string id))
+                    {
+                        relatedTo.Add(id);
+                    }
+                }
+
+                if (relatedTo.Count > 0)
+                {
+                    activity.AddTag(RelatedToTag, string.Join(",", relatedTo.Distinct()));
+                }
+            }
+        }
+
+        private Activity ProcessStart(string operationName, Message message, Func<object> getPayload, Action<Activity> setTags)
+        {
+            Activity activity = null;
+            string activityName = BaseActivityName + operationName;
+
+            if (DiagnosticListener.IsEnabled(activityName, entityPath))
+            {
+                var tmpActivity = message.ExtractActivity(activityName);
+                setTags?.Invoke(tmpActivity);
+                
+                if (DiagnosticListener.IsEnabled(activityName, entityPath, tmpActivity))
+                {
+                    activity = tmpActivity;
+                    if (DiagnosticListener.IsEnabled(activityName + ".Start"))
+                    {
+                        DiagnosticListener.StartActivity(activity, getPayload());
+                    }
+                    else
+                    {
+                        activity.Start();
+                    }
+                }
+            }
+            return activity;
+        }
+
+        private void SetTags(Activity activity, IList<Message> messageList)
+        {
+            var messageIds = messageList.Where(m => m.MessageId != null).Select(m => m.MessageId).ToArray();
+            if (messageIds.Any())
+            {
+                activity.AddTag(MessageIdTag, string.Join(",", messageIds));
+            }
+
+            var sessionIds = messageList.Where(m => m.SessionId != null).Select(m => m.SessionId).Distinct().ToArray();
+            if (sessionIds.Any())
+            {
+                activity.AddTag(SessionIdTag, string.Join(",", sessionIds));
+            }
+        }
+
+        private void SetTags(Activity activity, Message message)
```

Comment:
@lmolkova - shouldn't we set entity path and endpoint as tags here? these are very important for each monitoring solution so they should be made easy to consume

Reply:
as we discussed, we cannot leverage fully 'generic listener' and have to implement the custom listener for ServiceBus. The custom listener will be parsing payloads and extracting endpoint and entity from them. 
I wanted to keep instrumentation side as lightweight as possible. Otherwise, I would have to come up with 'standard' tags (as I'm sure similar things exists in almost every other Azure service client) and we would have to stick with them forever on AI side.